### PR TITLE
Adding NonMaxSuppression

### LINKED ISSUE #300: Unsupported shape calculation for operator nonMaximumSuppression
Hi everyone,

I'm trying to convert a mlmodel, which was created with turicreate Object Detection, to onnx. Unfortunately, I get the below issue:

`Unsupported shape calculation for operator nonMaximumSuppression`

### PR DESCRIPTION
**Pull request by vinitra**
As title. Resolves #300.

[NonMaxSuppression](https://github.com/apple/coremltools/blob/master/mlmodel/format/NonMaximumSuppression.proto) is implemented in CoreML as a model type as opposed to a neural network layer.

### CODE REVIEW

File: onnxmltools/convert/coreml/operator_converters/NonMaxSuppression.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+    if raw_model.HasField('iouThreshold'):
+        attrs['iou_threshold'] = np.array(raw_model.iouThreshold).astype(np.float32)
+    if raw_model.HasField('confidenceThreshold'):
+        attrs['score_threshold'] = np.array(raw_model.confidenceThreshold).astype(np.float32)
+
+    if raw_model.HasField('PickTop') and raw_model.PickTop.HasField('perClass'):
+        attrs['max_output_boxes_per_class'] = np.array(raw_model.PickTop.perClass).astype(np.float32)
+
+    container.add_node(op_type, [operator.inputs[0].full_name], [operator.outputs[0].full_name],
```
Comment (wschin) [inline onnxmltools/convert/coreml/operator_converters/NonMaxSuppression.py]: Could we throw if `targeted op_version < 10`?

File: onnxmltools/convert/coreml/operator_converters/neural_network/Activation.py [hunk 2]
```diff
@@ -46,6 +46,7 @@ def convert_activation(scope, operator, container):
     else:
         if activation_type == 'thresholdedReLU':
             op_type = 'ThresholdedRelu'
+            attrs['op_version'] = 10
```
Comment (wschin) [inline onnxmltools/convert/coreml/operator_converters/neural_network/Activation.py]: What if the targeted `op_version < 10`? Shouldn't it be another `apply_...`.

File: onnxmltools/convert/coreml/shape_calculators/NonMaxSuppression.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+from ...common._registration import register_shape_calculator
+from ...common.data_types import FloatTensorType, Int64TensorType
+from ...common.utils import check_input_and_output_numbers
+
+
+def calculate_non_max_suppression_output_shapes(operator):
+    '''
+    Allowed input/output patterns are
+        1.
```
Comment (wschin) [inline onnxmltools/convert/coreml/shape_calculators/NonMaxSuppression.py]: What is that?

File: onnxmltools/convert/coreml/operator_converters/neural_network/Activation.py (hunk 2 above)
Comment (vinitra) [reply onnxmltools/convert/coreml/operator_converters/neural_network/Activation.py]: ThresholdedReLu graduated to a full op from an experimental op -- it doesn't have an official implementation in the ONNX spec prior to Opset 10. Should we throw an error if container.target_opset < 10?

File: onnxmltools/convert/coreml/operator_converters/NonMaxSuppression.py (hunk 1 above)
Comment (vinitra) [reply onnxmltools/convert/coreml/operator_converters/NonMaxSuppression.py]: done!

File: onnxmltools/convert/coreml/shape_calculators/NonMaxSuppression.py (hunk 3 above)
Comment (vinitra) [reply onnxmltools/convert/coreml/shape_calculators/NonMaxSuppression.py]: Removed!

**Comment by pritasam** [comment]
Hi, I am still getting this error. Did it merged ??

**Comment by xadupre** [comment]
I don't see it master branch. It was not merged. It would help me if you had a simple example I could turn into a unit test.

**Comment by pritasam** [comment]
@xadupre  Hi I have one trained CoreML model. will it work if I share it with you ?

**Comment by pritasam** [comment]
Why merging is blocked ?

**Comment by wenbingl** [comment]
> Why merging is blocked ?

Need to some basic unit test to verify this code works.

**Comment by chrisballinger** [comment]
I'm also interested in seeing this merged. Could train a sample model and provide it, if that would help.

File: onnxmltools/convert/coreml/shape_calculators/NonMaxSuppression.py [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+# Licensed under the MIT License. See License.txt in the project root for
+# license information.
+# --------------------------------------------------------------------------
+
+from ...common._registration import register_shape_calculator
+from ...common.data_types import FloatTensorType, Int64TensorType
+from ...common.utils import check_input_and_output_numbers
+
+
+def calculate_non_max_suppression_output_shapes(operator):
```
Comment (chrisballinger) [inline onnxmltools/convert/coreml/shape_calculators/NonMaxSuppression.py:12]: @vinitra It seems to me that this function is missing an implementation, and is the missing piece of this PR. Do you have any guidance for the community on where to get started finishing this up?

Comment (vinitra) [reply onnxmltools/convert/coreml/shape_calculators/NonMaxSuppression.py:12]: Hi @chrisballinger! Thanks for your interest, and taking this on. I've addressed some of your review comments above. The remaining work:
- unit test
- shape inference calculations

I don't have the bandwidth for this currently, but would be happy to review PRs of anyone who wants to take this on.

Comment (chrisballinger) [reply onnxmltools/convert/coreml/shape_calculators/NonMaxSuppression.py:12]: Great! Do you have links to documentation or examples for where someone might get started? I found this for CoreML: https://apple.github.io/coremltools/coremlspecification/sections/FeatureEngineering.html#nonmaximumsuppression

Comment (vinitra) [reply onnxmltools/convert/coreml/shape_calculators/NonMaxSuppression.py:12]: Right, that's the link I was originally using from the CoreML side. There's also the original proto in coremltools, which has helpful comments:
https://github.com/apple/coremltools/blob/master/mlmodel/format/NonMaximumSuppression.proto
The relevant operator from Opset 11 on the ONNX side is here: 
https://github.com/onnx/onnx/blob/master/docs/Operators.md#NonMaxSuppression
For Opset 10, you can find the ONNX operator documentation below:
https://github.com/onnx/onnx/blob/master/docs/Changelog.md#NonMaxSuppression-10

Comment (chrisballinger) [reply onnxmltools/convert/coreml/shape_calculators/NonMaxSuppression.py:12]: This is super helpful, thank you so much!

**Comment by wenbingl** [comment]
It seems there is some additional work to complete this PR.
I temporarily close this PR until there is any new progress or update.

**Pull request closed without merging**